tsinfer/inference.py

```diff
@@ -490,6 +498,136 @@ def match_samples(
+    tree sequence.
+
+    Returns an ancestors tree sequence suitable for input into match_samples. If
+    ``create_sample_nodes`` is True, you probably want to exclude any noncontemporaneous
+    samples during the match_samples phase, otherwise they risk being duplicated in the
+    final tree sequence.
+    """
+
+    # First find the sites from the samples file used in the built ancestors
+    sites_used = np.isin(sample_data.sites_position[:], ancestor_data.sites_position[:])
+    sites_used_time = sample_data.sites_time[:][sites_used]
+    if np.any(sites_used_time == constants.TIME_UNSPECIFIED):
+        raise ValueError(
+            "Noncontemporaneous samples can only be ancestors if site times are present"
+        )
+    individuals_population = sample_data.individuals_population[:]
+
+    # Find the non-contemporaneous (nc) samples and group them by individuals_time
+    samples_individual = sample_data.samples_individual[:]
+    nc_individuals = sample_data.individuals_time[:] > 0
+    nc_samples = np.where(np.isin(samples_individual, np.where(nc_individuals)[0]))[0]
+    samples_by_timepoint = collections.defaultdict(list)
+    for s in nc_samples:
+        timepoint = sample_data.individuals_time[samples_individual[s]]
+        samples_by_timepoint[timepoint].append(s)
+
+    logger.info("Creating new ancestors file with added ancestors")
+    ancestor_to_sample_map = {}
+    inferred_ancestors = ancestor_data.ancestors()
+    current_anc = next(inferred_ancestors)
+    with formats.AncestorData(sample_data=sample_data) as new_data:
+        for insert_time, sample_ids in sorted(samples_by_timepoint.items()):
+            while current_anc is not None and current_anc.time > insert_time:
+                new_data.add_ancestor(**attr.asdict(current_anc, filter=exclude_id))
+                current_anc = next(inferred_ancestors, None)
+            for s in sample_ids:
+                logger.debug(f"Adding sample {s} as an ancestor")
+                haplotype = sample_data.sites_genotypes[:, s][sites_used]
+                if np.any(sites_used_time[haplotype > 0] < insert_time):
+                    bad = sites_used_time[haplotype > 0] < insert_time
+                    bad_pos = sample_data.sites_position[:][bad]
+                    raise ValueError(
+                        "Non-contemporanous ancestor is older than its derived mutations"
+                        + f" at sites {bad_pos}"
+                    )
+                anc_id = new_data.add_ancestor(
+                    start=0,
+                    end=ancestor_data.num_sites,
+                    time=insert_time,
+                    focal_sites=[],
+                    haplotype=haplotype,
+                )
+                ancestor_to_sample_map[anc_id] = s
+
+    logger.info("Matching ancestors including noncontemporaneous samples")
+    ancestor_tables = match_ancestors(
+        sample_data, ancestor_data, **kwargs
```

Comment:
I think this should be `match_ancestors(
         new_data, ancestor_data, **kwargs)`